pipelines/batch/src/main/java/org/openmrs/analytics/JdbcFetchHapi.java

```diff
@@ -83,54 +111,150 @@ public HapiRowDescriptor mapRow(ResultSet resultSet) throws Exception {
+        PCollection<ResourceTag> resourceTagPCollection) {
+      // Step to convert HapiRowDescriptor to key value struct <ResourceId, HapiRowDescriptor>,
+      // this will help in joining with tag PCollection.
+      PCollection<KV<String, HapiRowDescriptor>> resourceCollection =
+          hapiRowDescriptorPCollection.apply(
+              "Convert Resource Collection to KV",
+              ParDo.of(
+                  new DoFn<HapiRowDescriptor, KV<String, HapiRowDescriptor>>() {
+                    @ProcessElement
+                    public void processElement(ProcessContext processContext) {
+                      HapiRowDescriptor hapiRowDescriptor = processContext.element();
+                      processContext.output(
+                          KV.of(hapiRowDescriptor.resourceId(), hapiRowDescriptor));
+                    }
+                  }));
+
+      // Step to convert ResourceTag to key value struct <ResourceId, ResourceTag>,
+      // this will help in joining with resource PCollection.
+      PCollection<KV<String, ResourceTag>> tagCollection =
+          resourceTagPCollection.apply(
+              "Convert Tag Collection to KV",
+              ParDo.of(
+                  new DoFn<ResourceTag, KV<String, ResourceTag>>() {
+                    @ProcessElement
+                    public void processElement(ProcessContext processContext) {
+                      ResourceTag resourceTag = processContext.element();
+                      processContext.output(KV.of(resourceTag.getResourceId(), resourceTag));
+                    }
+                  }));
+
+      // Helper Tuples for joining.
+      TupleTag<HapiRowDescriptor> resourceTuple = new TupleTag<>();
+      TupleTag<ResourceTag> resourceTagTuple = new TupleTag<>();
+
+      // Join the collections.
+      PCollection<KV<String, CoGbkResult>> joinedResourceCollection =
+          KeyedPCollectionTuple.of(resourceTuple, resourceCollection)
+              .and(resourceTagTuple, tagCollection)
+              .apply(CoGroupByKey.create());
+
+      // Process the HapiRowDescriptor collection from the joined collection.
+      return joinedResourceCollection.apply(
+          "Join Resource and Tag Collections",
+          ParDo.of(
+              new DoFn<KV<String, CoGbkResult>, HapiRowDescriptor>() {
+                @ProcessElement
+                public void process(ProcessContext processContext) {
+                  KV<String, CoGbkResult> element = processContext.element();
+                  Iterable<HapiRowDescriptor> hapiRowDescriptorIterable =
+                      element.getValue().getAll(resourceTuple);
+                  Iterable<ResourceTag> resourceTagIterable =
+                      element.getValue().getAll(resourceTagTuple);
+
+                  List<ResourceTag> tags = new ArrayList<>();
+                  for (ResourceTag resourceTag : resourceTagIterable) {
+                    tags.add(resourceTag);
```

Comment:
Should the resourceTag be cloned as well?

Reply:
resourceTag is not being mutated